src/lightning/data/cache/serializers.py

```diff
@@ -0,0 +1,181 @@
+    """The PILSerializer serialize and deserialize PIL Image to and from bytes."""
+
+    def serialize(self, item: Image) -> bytes:
+        mode = item.mode.encode("utf-8")
+        width, height = item.size
+        raw = item.tobytes()
+        ints = np.array([width, height, len(mode)], np.uint32)
+        return ints.tobytes() + mode + raw
+
+    def deserialize(self, data: bytes) -> any:
+        idx = 3 * 4
+        width, height, mode_size = np.frombuffer(data[:idx], np.uint32)
+        idx2 = idx + mode_size
+        mode = data[idx:idx2].decode("utf-8")
+        size = width, height
+        raw = data[idx2:]
+        return Image.frombytes(mode, size, raw)  # pyright: ignore
+
+    def can_serialize(self, item) -> bool:
+        return isinstance(item, Image.Image) and not isinstance(item, JpegImageFile)
+
+
+class IntSerializer(Serializer):
+    """The IntSerializer serialize and deserialize integer to and from bytes."""
+
+    def serialize(self, item: int) -> bytes:
+        return str(item).encode("utf-8")
+
+    def deserialize(self, data: bytes) -> int:
+        return int(data.decode("utf-8"))
+
+    def can_serialize(self, item) -> bool:
+        return isinstance(item, int)
+
+
+class JPEGSerializer(Serializer):
+    """The JPEGSerializer serialize and deserialize JPEG image to and from bytes."""
+
+    def serialize(self, item: Image) -> bytes:
+        if isinstance(item, JpegImageFile):
+            if not hasattr(item, "filename"):
+                raise ValueError(
+                    "The JPEG Image's filename isn't defined. HINT: Open the image in your Dataset __getitem__ method."
+                )
+            with open(item.filename, "rb") as f:
+                return f.read()
+        raise TypeError(f"The provided itemect should be of type {JpegImageFile}. Found {item}.")
+
+    def deserialize(self, data: bytes) -> Image:
+        inp = BytesIO(data)
+        return Image.open(inp)
+
+    def can_serialize(self, item) -> bool:
+        return isinstance(item, JpegImageFile)
+
+
+class BytesSerializer(Serializer):
+    """The BytesSerializer serialize and deserialize integer to and from bytes."""
+
+    def serialize(self, item: bytes) -> bytes:
+        return item
+
+    def deserialize(self, item: bytes) -> bytes:
+        return item
+
+    def can_serialize(self, item: bytes) -> bool:
+        return isinstance(item, bytes)
+
+
+_TORCH_DTYPES_MAPPING = {
+    0: torch.float32,
+    1: torch.float,
+    2: torch.float64,
+    3: torch.double,
+    4: torch.complex64,
+    5: torch.cfloat,
+    6: torch.complex128,
+    7: torch.cdouble,
+    8: torch.float16,
+    9: torch.half,
+    10: torch.bfloat16,  # Not supported https://github.com/pytorch/pytorch/issues/110285
+    11: torch.uint8,
+    12: torch.int8,
+    13: torch.int16,
+    14: torch.short,
+    15: torch.int32,
+    16: torch.int,
+    17: torch.int64,
+    18: torch.long,
+    19: torch.bool,
+}
+
+
+class TensorSerializer(Serializer):
```

Comment:
Why can't we just use in-memory pickling? PyTorch already implements all of this.

Reply:
This removes the overhead of pickle. But I will benchmark pickle and checkout which is faster.

Reply:
Yes, you are right ! I think we should have it as a last resort.

Reply:
Quick benchmarking:

The custom TensorSerialization is roughly 4 times faster than pickle and generates files twice smaller. 

```python
def test_tensor_serializer():
    seed_everything(42)

    serializer_tensor = TensorSerializer()
    serializer_pickle = PickleSerializer()

    ratio_times = []
    ratio_bytes = []
    shapes = [(10,), (10, 10), (10, 10, 10), (10, 10, 10, 5), (10, 10, 10, 5, 4)]
    for dtype in _TORCH_DTYPES_MAPPING.values():
        for shape in shapes:
            # Not serializable for some reasons
            if dtype in [torch.bfloat16]:
                continue
            tensor = torch.ones(shape, dtype=dtype)

            t0 = time()
            data = serializer_tensor.serialize(tensor)
            deserialized_tensor = serializer_tensor.deserialize(data)
            tensor_time = time() - t0
            tensor_bytes = len(data)

            assert deserialized_tensor.dtype == dtype
            assert torch.equal(tensor, deserialized_tensor)

            t1 = time()
            data = serializer_pickle.serialize(tensor)
            deserialized_tensor = serializer_pickle.deserialize(data)
            pickle_time = time() - t1
            pickle_bytes = len(data)

            assert deserialized_tensor.dtype == dtype
            assert torch.equal(tensor, deserialized_tensor)

            ratio_times.append(pickle_time / tensor_time)
            ratio_bytes.append(pickle_bytes / tensor_bytes)

    assert np.mean(ratio_times) > 4
    assert np.mean(ratio_bytes) > 2
```